File: sdk/ts/src/output/formatters/text.ts
/**
 * @module output/formatters/text
 *
 * Built-in text formatter. Three styles: kv, lines, paragraph.
 *
 * - kv:        "key{sep}value" lines, blank line between records
 * - lines:     tab-separated, no header, one row per line
 * - paragraph: "Record N:" header line then "  field: value" lines
 *
 * Column order comes from the ColumnSpec list, or the first row's key order
 * when none was supplied; `cols` narrows and reorders.
 *
 * Zero deps. Mirrors Go text formatter byte-for-byte.
 */

import type { Formatter, Options } from '../formatter';
import { resolveColumnNames } from '../projection';

const STYLE_KV = 'kv';
const STYLE_LINES = 'lines';
const STYLE_PARAGRAPH = 'paragraph';

export const textFormatter: Formatter = {
  key: 'text',
  extensions: ['.txt'],
  options: [
    {
      name: 'style',
      type: 'enum',
      default: STYLE_KV,
      enum: [STYLE_KV, STYLE_LINES, STYLE_PARAGRAPH],
      usage: 'output style',
    },
    {
      name: 'separator',
      type: 'string',
      default: '=',
      usage: 'kv separator (kv style only)',
    },
  ],
  render(out, data, opts: Options, cols, columns) {
    const rows = normalise(data);
    // Emptiness is a ROW-count decision: no rows means no output at all.
    if (rows.length === 0) return;

    const headers = resolveColumnNames(rows, columns, cols);
    if (headers.length === 0) return;

    const style = ((opts['style'] as string) || STYLE_KV) as
      | 'kv'
      | 'lines'
      | 'paragraph';
    const sep = (opts['separator'] as string) || '=';

    switch (style) {
      case STYLE_KV:
        return renderKV(out, rows, headers, sep);
      case STYLE_LINES:
        return renderLines(out, rows, headers);
      case STYLE_PARAGRAPH:
        return renderParagraph(out, rows, headers);
    }
  },
};

function normalise(v: unknown): Record<string, unknown>[] {
  if (Array.isArray(v)) return v as Record<string, unknown>[];
  return [v as Record<string, unknown>];
}

function fmt(v: unknown): string {
  if (v === null || v === undefined) return '';
  return String(v);
}

function renderKV(
  out: NodeJS.WritableStream,
  rows: readonly Record<string, unknown>[],
  headers: readonly string[],
  sep: string,
): void {
  rows.forEach((r, i) => {
    if (i > 0) out.write('\n');
    for (const h of headers) {
      out.write(`${h}${sep}${fmt(r[h])}\n`);
    }
  });
}

function renderLines(
  out: NodeJS.WritableStream,
  rows: readonly Record<string, unknown>[],
  headers: readonly string[],
): void {
  for (const r of rows) {
    out.write(headers.map(h => fmt(r[h])).join('\t') + '\n');
  }
}

function renderParagraph(
  out: NodeJS.WritableStream,
  rows: readonly Record<string, unknown>[],
  headers: readonly string[],
): void {
  rows.forEach((r, i) => {
    if (i > 0) out.write('\n');
    out.write(`Record ${i + 1}:\n`);
    for (const h of headers) {
      out.write(`  ${h}: ${fmt(r[h])}\n`);
    }
  });
}

File: sdk/experimental/rs/src/output/builtins/text.rs
//! Plain-text formatter — kv / lines / paragraph styles.
//!
//! Mirrors go `console/output/text.go` byte-for-byte, which py and ts also
//! reproduce exactly. No library is involved or wanted: the three styles are
//! trivial renderers over an ordered column list, and matching the existing
//! implementations exactly is the entire job.
//!
//! - `style=kv` (default): `HEADER<sep>VALUE\n` per field, blank line
//!   BETWEEN records (never trailing).
//! - `style=lines`: values tab-joined, one record per line, no header.
//! - `style=paragraph`: `Record N:\n` then `  HEADER: VALUE\n` lines, blank
//!   line BETWEEN records. N is 1-indexed.
//!
//! Options:
//!   - `style` (enum kv|lines|paragraph, default kv) — output style.
//!   - `separator` (string, default "=") — kv separator (kv style only).

use std::io::Write;

use serde_json::Value;

use crate::output::formatter::Formatter;
use crate::output::option::{OptionSpec, OptionType, OptionValue, Options};

use super::rows::{normalize, resolve_columns, row_get, stringify};

const STYLE_KV: &str = "kv";
const STYLE_LINES: &str = "lines";
const STYLE_PARAGRAPH: &str = "paragraph";

static OPTS: &[OptionSpec] = &[
    OptionSpec {
        name: "style",
        r#type: OptionType::Enum,
        usage: "output style",
        default: None, // merged at render-time
        r#enum: &[STYLE_KV, STYLE_LINES, STYLE_PARAGRAPH],
    },
    OptionSpec {
        name: "separator",
        r#type: OptionType::String,
        usage: "kv separator (kv style only)",
        default: None,
        r#enum: &[],
    },
];

pub struct TextFormatter;

impl Formatter for TextFormatter {
    fn key(&self) -> &'static str {
        "text"
    }

    fn extensions(&self) -> &'static [&'static str] {
        &[".txt"]
    }

    fn options(&self) -> &'static [OptionSpec] {
        OPTS
    }

    fn render(
        &self,
        out: &mut dyn Write,
        data: &Value,
        opts: &Options,
        cols: &[String],
    ) -> std::io::Result<()> {
        let rows = normalize(data);

        // Zero rows emits nothing. Guarded on ROW count, never on column
        // count — `cols` is populated for row-less payloads too.
        if rows.is_empty() {
            return Ok(());
        }

        let columns = resolve_columns(&rows, cols);

        let style = opts
            .get("style")
            .and_then(OptionValue::as_str)
            .filter(|s| !s.is_empty())
            .unwrap_or(STYLE_KV);

        match style {
            STYLE_KV => {
                let sep = opts
                    .get("separator")
                    .and_then(OptionValue::as_str)
                    .filter(|s| !s.is_empty())
                    .unwrap_or("=");
                render_kv(out, &rows, &columns, sep)
            }
            STYLE_LINES => render_lines(out, &rows, &columns),
            STYLE_PARAGRAPH => render_paragraph(out, &rows, &columns),
            other => Err(std::io::Error::new(
                std::io::ErrorKind::InvalidInput,
                format!("text formatter: unknown style {other:?}"),
            )),
        }
    }
}

fn render_kv(
    out: &mut dyn Write,
    rows: &[&Value],
    columns: &[String],
    sep: &str,
) -> std::io::Result<()> {
    for (i, row) in rows.iter().enumerate() {
        if i > 0 {
            out.write_all(b"\n")?;
        }
        for c in columns {
            let line = format!("{}{}{}\n", c, sep, stringify(row_get(row, c)));
            out.write_all(line.as_bytes())?;
        }
    }
    Ok(())
}

fn render_lines(out: &mut dyn Write, rows: &[&Value], columns: &[String]) -> std::io::Result<()> {
    for row in rows {
        let cells: Vec<String> = columns.iter().map(|c| stringify(row_get(row, c))).collect();
        out.write_all(cells.join("\t").as_bytes())?;
        out.write_all(b"\n")?;
    }
    Ok(())
}

fn render_paragraph(
    out: &mut dyn Write,
    rows: &[&Value],
    columns: &[String],
) -> std::io::Result<()> {
    for (i, row) in rows.iter().enumerate() {
        if i > 0 {
            out.write_all(b"\n")?;
        }
        out.write_all(format!("Record {}:\n", i + 1).as_bytes())?;
        for c in columns {
            let line = format!("  {}: {}\n", c, stringify(row_get(row, c)));
            out.write_all(line.as_bytes())?;
        }
    }
    Ok(())
}
